File: apps/python/event_sourcing/event.py
"""Event base class and concrete events."""

from abc import ABC
from datetime import datetime
from typing import Union


class Event(ABC):
    """Base class for all events."""

    def __init__(self, recorded_at: datetime):
        self.type: str = ""
        self.recorded_at = recorded_at


class RestockOrdered(Event):
    """Event emitted when a restock order is placed."""

    TYPE: str = "restock_ordered"

    def __init__(self, quantity: int):
        super().__init__(datetime.now())
        self.type = RestockOrdered.TYPE
        self.quantity = quantity


# Type alias for all event types
EVENTS = Union[RestockOrdered]

File: apps/rust/src/event.rs
//! Event types and trait definition.

use chrono::{DateTime, Utc};

/// Base trait for all events.
pub trait Event {
    /// Get the event type identifier.
    fn event_type(&self) -> &str;

    /// Get the timestamp when the event was recorded.
    fn recorded_at(&self) -> DateTime<Utc>;
}

/// Event emitted when a restock order is placed.
#[derive(Debug, Clone, PartialEq)]
pub struct RestockOrdered {
    pub quantity: i32,
    pub recorded_at: DateTime<Utc>,
}

impl RestockOrdered {
    pub const TYPE: &'static str = "restock_ordered";

    pub fn new(quantity: i32) -> Self {
        Self {
            quantity,
            recorded_at: Utc::now(),
        }
    }
}

impl Event for RestockOrdered {
    fn event_type(&self) -> &str {
        Self::TYPE
    }

    fn recorded_at(&self) -> DateTime<Utc> {
        self.recorded_at
    }
}
